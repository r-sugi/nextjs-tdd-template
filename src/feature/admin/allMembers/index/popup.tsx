import { type PropsWithChildren, useState } from "react";

export const PopupMenu = ({ children }: PropsWithChildren) => {
	const [isDropdownVisible, setDropdownVisible] = useState(false);

	const handleFocus = () => {
		setDropdownVisible(true);
	};

	const handleBlur = () => {
		setDropdownVisible(false);
	};

	return (
		<nav className="flex items-center justify-between h-full p-3 m-auto">
			<div className="relative">
				<button
					id="user-menu"
					type="button"
					aria-haspopup="true"
					onFocus={handleFocus}
					onBlur={handleBlur}
				>
					{children}
				</button>

				<div
					id="user-menu-dropdown"
					className={`z-10 absolute right-0 w-48 mt-2 origin-top-right rounded-lg shadow-lg transform ${
						isDropdownVisible
							? "scale-100 opacity-100 ease-in duration-100"
							: "scale-0 opacity-0 ease-out duration-75"
					} top-13`}
				>
					<div
						className="p-4 bg-white rounded-md shadow-xs"
						role="menu"
						aria-orientation="vertical"
						aria-labelledby="user-menu"
					>
						<div
							className="block px-6 py-2 mb-2 font-bold rounded"
							role="menuitem"
						>
							Delete
						</div>
					</div>
				</div>
			</div>
		</nav>
	);
};
