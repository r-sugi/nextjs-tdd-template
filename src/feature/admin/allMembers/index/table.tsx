import type { MembersByType } from "@/core/domains/member/member";
import { ThreeDotsIcon } from "./three-dots.icon";

type PropsType = {
	members: MembersByType;
};

export const MemberTable = ({ members }: PropsType) => {
	const tableHeaders = [
		"Created At",
		// TODO: ドメインモデルの定義、ドメインロジックの定義を修正する
		// "Email",
		// "Member ID",
		"Status Activity ID",
		"Type",
		"Status",
		"",
	];

	return (
		<table className="min-w-full bg-white border border-gray-300">
			<thead>
				<tr>
					{tableHeaders.map((header) => (
						<th
							key={header}
							className="py-2 px-4 bg-gray-200 border-b border-gray-300"
						>
							{header}
						</th>
					))}
				</tr>
			</thead>
			<tbody>
				{members.map((member) => (
					<tr key={member.statusActivityId}>
						<td className="py-2 px-4 border-b border-gray-300">
							{member.createdAt.getTime()}
						</td>
						{/* // TODO: ドメインモデルの定義、ドメインロジックの定義を修正する */}
						{/* <td className="py-2 px-4 border-b border-gray-300">
							{member.email}
						</td> */}
						{/* <td className="py-2 px-4 border-b border-gray-300">
							{member.memberId}
						</td> */}
						<td className="py-2 px-4 border-b border-gray-300">
							{member.statusActivityId}
						</td>
						<td className="py-2 px-4 border-b border-gray-300">
							{member.status}
						</td>
						<td className="py-2 px-4 border-b border-gray-300">
							{member.status}
						</td>
						<td className="cursor-pointer py-2 px-4 border-b border-gray-300">
							<div className="flex justify-between">
								<div className="hover:text-blue-500 hover:underline">
									Details
								</div>
								<div className="items-center">
									<ThreeDotsIcon />
								</div>
							</div>
						</td>
					</tr>
				))}
			</tbody>
		</table>
	);
};
