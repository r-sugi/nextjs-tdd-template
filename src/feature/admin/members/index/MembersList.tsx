import type { MembersByType } from "@/core/domains/member/member";
import type { FC } from "react";

type Props = {
	members: MembersByType;
};

export const MembersList: FC<Props> = (props) => {
	if (props.members.length === 0) {
		return <div>メンバーがいません</div>;
	}

	return (
		<ul>
			{props.members.map((member) => (
				<div className="text-xl font-bold" key={member.statusActivityId}>
					{JSON.stringify(member, null, 2)}
				</div>
			))}
		</ul>
	);
};
