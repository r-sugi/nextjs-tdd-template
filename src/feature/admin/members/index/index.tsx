import { FC } from 'react';

import { memberStatus } from '@/core/domains/member/status';
import { useFetchMembers } from '@/core/usecases/member/useFetchMembers.query';

export const IndexTemplate: FC = () => {
  const { data, refetch, loading } = useFetchMembers();

  if (loading) {
    return <div>loading...</div>;
  }

  const memberListComponent = () => {
    return (
      <ul>
        {data.members.map((member) => (
          <li key={member.statusActivityId}>{JSON.stringify(member, null, 2)}</li>
        ))}
      </ul>
    );
  };

  return (
    <div>
      {/* TODO: styleに関するリファクタ */}
      <style jsx={true}>{`
        .tab-wrap {
          display: flex;
          flex-wrap: wrap;
        }
        .tab-label {
          color: White;
          background: LightGray;
          margin-right: 5px;
          padding: 3px 12px;
          order: -1;
        }
        .tab-content {
          width: 100%;
          display: none;
        }
        /* アクティブなタブ */
        .tab-switch:checked + .tab-label {
          background: DeepSkyBlue;
        }
        .tab-switch:checked + .tab-label + .tab-content {
          display: block;
        }
        /* ラジオボタン非表示 */
        .tab-switch {
          display: none;
        }
      `}</style>
      {/* TODO: componentsのリファクタ */}
      {/* メンバー一覧 */}
      <div className="tab-wrap">
        <input
          id={memberStatus.pendingActivation}
          type="radio"
          name="TAB"
          className="tab-switch"
          checked={data.queryMemberStatus === memberStatus.pendingActivation}
          onChange={() => refetch(memberStatus.pendingActivation)}
        />
        <label className="tab-label" htmlFor={memberStatus.pendingActivation}>
          認証前
        </label>
        <div className="tab-content">{memberListComponent()}</div>
        <input
          id={memberStatus.active}
          type="radio"
          name="TAB"
          className="tab-switch"
          checked={data.queryMemberStatus === memberStatus.active}
          onChange={() => refetch(memberStatus.active)}
        />
        <label className="tab-label" htmlFor={memberStatus.active}>
          アクティブ
        </label>
        <div className="tab-content">{memberListComponent()}</div>
        <input
          id={memberStatus.resigned}
          type="radio"
          name="TAB"
          className="tab-switch"
          checked={data.queryMemberStatus === memberStatus.resigned}
          onChange={() => refetch(memberStatus.resigned)}
        />
        <label className="tab-label" htmlFor={memberStatus.resigned}>
          退会済み
        </label>
        <div className="tab-content">{memberListComponent()}</div>
        <input
          id={memberStatus.banned}
          type="radio"
          name="TAB"
          className="tab-switch"
          checked={data.queryMemberStatus === memberStatus.banned}
          onChange={() => refetch(memberStatus.banned)}
        />
        <label className="tab-label" htmlFor={memberStatus.banned}>
          アカウント停止
        </label>
        <div className="tab-content">{memberListComponent()}</div>
      </div>
    </div>
  );
};