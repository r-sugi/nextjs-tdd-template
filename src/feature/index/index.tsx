import { FC } from 'react';
export const IndexTemplate: FC = () => {
  return <div>トップページです</div>;
};
