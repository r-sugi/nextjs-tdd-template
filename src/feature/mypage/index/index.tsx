import type { FC } from "react";

import { apolloErrorHandler } from "@/core/repositories/apolloErrorHandler";
import { useFetchActiveMember } from "@/core/usecases/member/useFetchActiveMember.query";

const IndexTemplate: FC = () => {
	const { data: activeMember, loading, error } = useFetchActiveMember();

	if (loading) {
		return <div>loading...</div>;
	}

	if (error) {
		const graphQLErrors = apolloErrorHandler(error);
		return <div>graphQLErrors: {JSON.stringify(graphQLErrors, null, 2)}</div>;
	}

	if (activeMember == null) {
		// メンバーが存在しない場合はリダイレクト
	}

	return <div>ActiveMember: {JSON.stringify(activeMember, null, 2)}</div>;
};

export default IndexTemplate;
