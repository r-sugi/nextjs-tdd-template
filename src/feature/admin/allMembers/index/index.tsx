import type { FC } from "react";

import { useFetchAllMembers } from "@/core/usecases/member/useFetchAllMembers.query";

import { MemberTableRow } from "./table/MemberTableRow";
import { MemberTable } from "./table/table";
import { type OnSubmitStatusChange, eventTypes } from "./table/type";

export const IndexTemplate: FC = () => {
	const { data, loading } = useFetchAllMembers();

	if (loading) {
		return <div>loading...</div>;
	}
	if (data?.members === null) {
		return (
			<div data-testid="admin-members-index-error">
				CSR リクエストエラー: データ取得に失敗しました
			</div>
		);
	}

	// Event
	const onSubmit: OnSubmitStatusChange = (payload) => {
		if (payload.type === eventTypes.onClickBan) {
			console.log("API処理", payload.detail);
		} else if (payload.type === eventTypes.onClickDisable) {
			console.log("API処理", payload.detail);
		}
	};

	return (
		<div data-testid="admin-members-index">
			{/* メンバー一覧 */}
			<div className="container mx-auto">
				<h1 className="text-2xl font-bold mb-4">Members</h1>
				<MemberTable>
					{data.members.map((member) => (
						<MemberTableRow
							member={member}
							key={member.statusActivityId}
							onSubmit={onSubmit}
						/>
					))}
				</MemberTable>
			</div>
		</div>
	);
};
