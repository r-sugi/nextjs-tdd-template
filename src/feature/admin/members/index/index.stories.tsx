import { GetMembersByStatusDocument } from "@/generated/graphql";
import type { Meta, StoryObj } from "@storybook/react";
import { screen } from "@storybook/testing-library";

import { graphql } from "msw";

import { fixtureGetMembersByStatus } from "mocks/fixtures/member";

import { IndexTemplate } from "./index";
import { StubAuthProvider } from "@/feature/auth/provider/StubAuthProvider";
import { stubAuthContext } from "mocks/fixtures/provider/useStubAuthContext";

export default {
	component: IndexTemplate,
	parameters: {
		// see https://storybook.js.org/docs/react/configure/story-layout
		layout: "padded", // available "fullscreen" and "centered"
	},
} as Meta<typeof IndexTemplate>;

type Story = StoryObj<typeof IndexTemplate>;

export const タブ切替: Story = {
	decorators: [
		(Story) => (
			<StubAuthProvider value={stubAuthContext.signedIn}>
				<Story />
			</StubAuthProvider>
		),
	],
	parameters: {
		screenshot: {
			waitFor: async () => {
				await screen.findByText("statusActivityId");
			},
		},
		msw: {
			handlers: [
				graphql.query(GetMembersByStatusDocument, (_, res, ctx) => {
					return res(ctx.data(fixtureGetMembersByStatus.data), ctx.delay(100));
				}),
			],
		},
	},
};
