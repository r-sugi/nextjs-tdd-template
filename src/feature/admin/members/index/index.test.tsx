import { render, waitFor, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import {
	GetMembersByStatusDocument,
	ResignMemberDocument,
} from "@/generated/graphql";
import { AppProvider } from "@/pages/_provider/_app.provider";
import { fixtureGetMembersByStatus } from "@/mocks/fixtures/member";
import { HttpResponse, graphql } from "msw";
import { setupServer } from "msw/node";
import { IndexTemplate } from "./index";
jest.mock("@/utils/cache");

const server = setupServer();
beforeAll(() => server.listen({ onUnhandledRequest: "bypass" }));
afterEach(() => {
	jest.resetAllMocks();
	jest.restoreAllMocks();
	server.resetHandlers();
});
afterAll(() => server.close());

describe(IndexTemplate, () => {
	const setup = () => {
		return {
			user: userEvent.setup(),
			view: render(
				<AppProvider>
					<IndexTemplate />
				</AppProvider>,
			),
		};
	};

	describe("正常系", () => {
		it("初期描画が変化していない", async () => {
			const mockFn = jest.fn();
			server.use(
				graphql.query(GetMembersByStatusDocument, ({ variables }) => {
					mockFn(variables);
					return HttpResponse.json({
						data: fixtureGetMembersByStatus.data,
					});
				}),
			);

			const { view } = setup();
			const viewRoot = await view.findByTestId("admin-members-index");
			expect(viewRoot).toBeVisible();
			expect(view.container).toMatchSnapshot();
		});
	});

	describe("異常系", () => {
		it("APIエラーを表示する", async () => {
			// Arrange
			const mockFn = jest.fn();
			server.use(
				graphql.query(GetMembersByStatusDocument, ({ variables }) => {
					mockFn(variables);
					return HttpResponse.json({
						errors: [],
					});
				}),
			);
			// Act
			const { view } = setup();

			// Assert
			const viewRoot = await view.findByTestId("admin-members-index-error");
			const errorEl = within(viewRoot).getByText(
				"CSR リクエストエラー: データ取得に失敗しました",
			);
			expect(errorEl).toBeVisible();
		});
	});
});
