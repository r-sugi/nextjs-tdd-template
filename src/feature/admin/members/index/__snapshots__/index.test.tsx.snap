// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexTemplate 正常系 初期描画が変化していない 1`] = `
<div>
  <div
    class="jsx-6f8c5883fcd245f"
    data-testid="admin-members-index"
  >
    <div
      class="jsx-6f8c5883fcd245f tab-wrap"
    >
      <input
        checked=""
        class="jsx-6f8c5883fcd245f tab-switch"
        id="pendingActivation"
        name="TAB"
        type="radio"
      />
      <label
        class="jsx-6f8c5883fcd245f tab-label"
        for="pendingActivation"
      >
        認証前
      </label>
      <div
        class="jsx-6f8c5883fcd245f tab-content"
      >
        <ul>
          <li>
            {
  "createdAt": "2024-04-15T11:34:46.135Z",
  "email": "foo@sample.com",
  "memberId": "5422c606-4cca-4054-ac66-83b682c53584",
  "statusActivityId": "128f3ea5-962b-4119-9d5c-05084b103708",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
          <li>
            {
  "createdAt": "2024-04-21T10:31:43.791Z",
  "email": "hhhh@gmail.com",
  "memberId": "f0f04501-dd5d-4418-9e2f-c3b0b62941f2",
  "statusActivityId": "32833818-04fd-458e-a016-661bec557de6",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
        </ul>
      </div>
      <input
        class="jsx-6f8c5883fcd245f tab-switch"
        id="active"
        name="TAB"
        type="radio"
      />
      <label
        class="jsx-6f8c5883fcd245f tab-label"
        for="active"
      >
        アクティブ
      </label>
      <div
        class="jsx-6f8c5883fcd245f tab-content"
      >
        <ul>
          <li>
            {
  "createdAt": "2024-04-15T11:34:46.135Z",
  "email": "foo@sample.com",
  "memberId": "5422c606-4cca-4054-ac66-83b682c53584",
  "statusActivityId": "128f3ea5-962b-4119-9d5c-05084b103708",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
          <li>
            {
  "createdAt": "2024-04-21T10:31:43.791Z",
  "email": "hhhh@gmail.com",
  "memberId": "f0f04501-dd5d-4418-9e2f-c3b0b62941f2",
  "statusActivityId": "32833818-04fd-458e-a016-661bec557de6",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
        </ul>
      </div>
      <input
        class="jsx-6f8c5883fcd245f tab-switch"
        id="resigned"
        name="TAB"
        type="radio"
      />
      <label
        class="jsx-6f8c5883fcd245f tab-label"
        for="resigned"
      >
        退会済み
      </label>
      <div
        class="jsx-6f8c5883fcd245f tab-content"
      >
        <ul>
          <li>
            {
  "createdAt": "2024-04-15T11:34:46.135Z",
  "email": "foo@sample.com",
  "memberId": "5422c606-4cca-4054-ac66-83b682c53584",
  "statusActivityId": "128f3ea5-962b-4119-9d5c-05084b103708",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
          <li>
            {
  "createdAt": "2024-04-21T10:31:43.791Z",
  "email": "hhhh@gmail.com",
  "memberId": "f0f04501-dd5d-4418-9e2f-c3b0b62941f2",
  "statusActivityId": "32833818-04fd-458e-a016-661bec557de6",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
        </ul>
      </div>
      <input
        class="jsx-6f8c5883fcd245f tab-switch"
        id="banned"
        name="TAB"
        type="radio"
      />
      <label
        class="jsx-6f8c5883fcd245f tab-label"
        for="banned"
      >
        アカウント停止
      </label>
      <div
        class="jsx-6f8c5883fcd245f tab-content"
      >
        <ul>
          <li>
            {
  "createdAt": "2024-04-15T11:34:46.135Z",
  "email": "foo@sample.com",
  "memberId": "5422c606-4cca-4054-ac66-83b682c53584",
  "statusActivityId": "128f3ea5-962b-4119-9d5c-05084b103708",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
          <li>
            {
  "createdAt": "2024-04-21T10:31:43.791Z",
  "email": "hhhh@gmail.com",
  "memberId": "f0f04501-dd5d-4418-9e2f-c3b0b62941f2",
  "statusActivityId": "32833818-04fd-458e-a016-661bec557de6",
  "__typename": "memberPendingActivation",
  "status": "pendingActivation"
}
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
